stats/utils/src/main/java/org/apache/bookkeeper/stats/utils/StatsDocGenerator.java

```diff
@@ -0,0 +1,302 @@
+        StatsDocEntry docEntry = new StatsDocEntry(statsName, statsType, helpDesc);
+        categoryMap.put(statsName, docEntry);
+    }
+
+    private void writeDoc(String file) throws IOException {
+        DumperOptions options = new DumperOptions();
+        options.setDefaultFlowStyle(FlowStyle.BLOCK);
+        options.setDefaultScalarStyle(ScalarStyle.LITERAL);
+        Yaml yaml = new Yaml(options);
+        Writer writer;
+        if (Strings.isNullOrEmpty(file)) {
+            writer = new StringWriter();
+        } else {
+            writer = new OutputStreamWriter(new FileOutputStream(file), UTF_8);
+        }
+        try {
+            Map<String, Map<String, Map<String, String>>> docs = docEntries.entrySet()
+                .stream()
+                .collect(Collectors.toMap(
+                    e -> e.getKey(),
+                    e -> e.getValue().entrySet()
+                        .stream()
+                        .collect(Collectors.toMap(
+                            e1 -> e1.getKey(),
+                            e1 -> e1.getValue().properties()
+                        ))
+                ));
+            yaml.dump(docs, writer);
+            writer.flush();
+            if (writer instanceof StringWriter) {
+                StringWriter sw = (StringWriter) writer;
+                System.out.println(sw.toString());
+            }
+        } finally {
+            writer.close();
+        }
+    }
+
+    /**
+     * Args for stats generator.
+     */
+    private static class MainArgs {
+
+        @Parameter(
+            names = {
+                "-p", "--package"
+            },
+            description = "Package prefix of the classes to generate stats doc")
+        String packagePrefix = "org.apache.bookkeeper";
+
+        @Parameter(
+            names = {
+                "-sp", "--stats-provider"
+            },
+            description = "The stats provider used for generating stats doc")
+        String statsProviderClass = "prometheus";
+
+        @Parameter(
+            names = {
+                "-o", "--output-file"
+            },
+            description = "The output file to dump stats docs"
+        )
+        String yamlFile = null;
+
+        @Parameter(
+            names = {
+                "-h", "--help"
+            },
+            description = "Show this help message")
+        boolean help = false;
+
+    }
+
+    public static void main(String[] args) throws Exception {
+        MainArgs mainArgs = new MainArgs();
+
+        JCommander commander = new JCommander();
+        try {
+            commander.setProgramName("stats-doc-gen");
+            commander.addObject(mainArgs);
+            commander.parse(args);
+            if (mainArgs.help) {
+                commander.usage();
+                Runtime.getRuntime().exit(0);
+                return;
+            }
+        } catch (Exception e) {
+            commander.usage();
+            Runtime.getRuntime().exit(-1);
```

Comment:
here is java function. per javadoc, a nonzero status abnormal termination. it doesn't prohibit negative values. as it is running on unix-ish platform, bash will probably scope it into 8 bits. so it will be showed up as 255.

Reply:
It doesn't prohibit, but standard practice is 0-255.

Reply:
it is pretty common at java returning -1. the standard practice is posix, but not java. I would rather not enforcing non-java standard in java practice if it is not strongly required.

Reply:
@ivankelly this is a tool. we don't really use the exit code here. what we care whether it is a success or a failure. there is no point for defining an exit code for it and jvm already handles that.